### ISSUE TITLE
PHP namespaces adoption

### ISSUE THREAD

**Issue by thekid** [xp-framework/rfc#136]
# Scope of Change

This strategy describes the PHP namespaces adoption strategy.
# Rationale

Once and for all remove the need to prefix classes.
# Functionality

PHP Namespace syntax:

``` php
<?php namespace net\xp_framework\example;

use util\Date;

class ExampleClass extends \lang\Object {
}
```
## Step 1: Support

See RFC #222 - optional support for namespaces and enable forward-compatibility for third-party libraries. Users can now use fully qualified names for XP classes or unqualified names as before:

``` sh
$ xp -w 'new \util\Date();'
2012-02-22 00:46:57+0100

$ xp -w 'uses("util.Date"); new Date();'
2012-02-22 00:47:17+0100
```

In both cases, the _real_ class name (see `XPClass::literal()` method) is `Date` though, the qualified version is created using `class_alias()`.
## Step 2: Migration

Migrate framework classes to PHP namespaces, deprecate RFC #37 fully qualified classes - XP 6.0.0.

Users can still use both fully qualified and unqualified names for XP classes, although the _real_ name now is `util\Date`, and the short name is created using `class_alias()`.

``` sh
$ xp -w 'new \util\Date();'
2012-02-22 00:46:57+0100

$ xp -w 'uses("util.Date"); new Date();'
2012-02-22 00:47:17+0100
```
## Step 3: Cleanup

Remove BC measures and RFC #37 fully qualified classes. No longer support non-namespaced classes - XP 6.1.0. 

Using the non-qualified names is no longer supported.
# Security considerations

n/a
# Speed impact

n/a
# Dependencies
- PHP with namespaces support (5.3.0)
# Related documents
- http://news.xp-framework.net/article/206/0000/00/00/ - Namespaces briefing
- RFC #172 - XP 6

**Comment by thekid** [comment]
# For Step 2 (XP 6.0.0)

First working, though buggy version, committed in thekid/xp-framework@cabde6b - generated by:

``` sh
$ xpcli -cp ../xp.experiments/ns-convert/ ConvertToNamespaces core/src/main/php ../xp.thekid/core/src/main/php/ lang.SystemExit scriptlet.HttpScriptletException unittest.TestCase peer.inet.InetAddress
$ xpcli -cp ../xp.experiments/ns-convert/ ConvertToNamespaces tools/src/main/php ../xp.thekid/tools/src/main/php/ lang.SystemExit scriptlet.HttpScriptletException unittest.TestCase
```

...from a checkout with the `xp5_9` branch and _without_ the directory `core/src/main/php/text/format/`, overlayed with:

``` diff
--- a/core/src/main/php/lang/ClassLoader.class.php
+++ b/core/src/main/php/lang/ClassLoader.class.php
@@ -160,10 +160,18 @@
           }
         }

+        if (FALSE === ($p= strrpos($name, '\\'))) {
+          $decl= $name;
+          $namespace= NULL;
+        } else {
+          $decl= substr($name, $p+ 1);
+          $namespace= substr($name, 0, $p);
+        }
         with ($dyn= self::registerLoader(DynamicClassLoader::instanceFor(__METHOD__))); {
           $dyn->setClassBytes($class, sprintf(
-            'class %s extends %s%s %s',
-            $name,
+            '%sclass %s extends %s%s %s',
+            $namespace ? 'namespace '.$namespace.'; ' : '',
+            $decl,
             $super,
             $interfaces ? ' implements '.implode(', ', $if) : '',
             $bytes
@@ -197,10 +205,18 @@
           }
         }

+        if (FALSE === ($p= strrpos($name, '\\'))) {
+          $decl= $name;
+          $namespace= NULL;
+        } else {
+          $decl= substr($name, $p+ 1);
+          $namespace= substr($name, 0, $p);
+        }
         with ($dyn= self::registerLoader(DynamicClassLoader::instanceFor(__METHOD__))); {
           $dyn->setClassBytes($class, sprintf(
-            'interface %s%s %s',
-            $name,
+            '%sinterface %s%s %s',
+            $namespace ? 'namespace '.$namespace.'; ' : '',
+            $decl,
             sizeof($if) ? ' extends '.implode(', ', $if) : '',
             $bytes
           ));
```

## Proof of concept

``` sh
Timm Friebe@samson ~/devel/xp.thekid [namespaces]
$ USE_XP="core;tools" xp -v
XP 6.0.0-dev { PHP 5.3.10 & ZE 2.3.0 } @ Windows NT SAMSON 6.1 build 7600 (Windows 7 Home Premium Edition) i586
Copyright (c) 2001-2012 the XP group
lang.FileSystemClassLoader<D:\cygwin\home\Timm Friebe\devel\xp.thekid\core\src\main\php\>
lang.FileSystemClassLoader<D:\cygwin\home\Timm Friebe\devel\xp.thekid\core\src\test\php\>
lang.FileSystemClassLoader<D:\cygwin\home\Timm Friebe\devel\xp.thekid\tools\src\main\php\>
lang.FileSystemClassLoader<D:\cygwin\home\Timm Friebe\devel\xp.thekid\>
```

## Syntax example

``` php
<?php namespace lang;

/**
 * Indicates a method was tried to be called or a property was tried
 * to be accessed on a non-object.
 *
 * @test  xp://net.xp_framework.unittest.core.NullTest
 */
class NullPointerException extends XPException {

}
```

## Success story

``` sh
$ USE_XP="core;tools" unittest net.xp_framework.unittest.core.ObjectTest
[...........]

OK: 11/11 run (0 skipped), 11 succeeded, 0 failed
Memory used: 2322.06 kB (2583.48 kB peak)
Time taken: 0.001 seconds
```

**Comment by thekid** [comment]
- [x] Step 2 now scheduled in xp-framework/rfc#172

**Comment by thekid** [comment]
Implemented with xp-framework/core@f6c19d4 - Full diff (for the patient): https://github.com/xp-framework/core/compare/a160bfe...f6c19d4
